File: typescript/src/liveness.ts
export interface LivenessSweep<Connection> {
  /** Connections that were alive and should receive the next probe. */
  readonly probe: readonly Connection[];
  /** Connections that did not acknowledge the previous probe. */
  readonly stale: readonly Connection[];
}

/** Pure acknowledgement state for heartbeat adapters. */
export class LivenessTracker<Connection extends object> {
  readonly #alive = new Map<Connection, boolean>();

  track(connection: Connection): boolean {
    if (this.#alive.has(connection)) return false;
    this.#alive.set(connection, true);
    return true;
  }

  acknowledge(connection: Connection): boolean {
    if (!this.#alive.has(connection)) return false;
    this.#alive.set(connection, true);
    return true;
  }

  remove(connection: Connection): boolean {
    return this.#alive.delete(connection);
  }

  sweep(): LivenessSweep<Connection> {
    const probe: Connection[] = [];
    const stale: Connection[] = [];
    for (const [connection, alive] of this.#alive) {
      if (alive) {
        probe.push(connection);
        this.#alive.set(connection, false);
      } else {
        stale.push(connection);
        this.#alive.delete(connection);
      }
    }
    return { probe, stale };
  }

  size(): number {
    return this.#alive.size;
  }

  clear(): number {
    const removed = this.#alive.size;
    this.#alive.clear();
    return removed;
  }
}

File: rust/src/liveness.rs
use std::collections::BTreeMap;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct LivenessSweep<Connection> {
    pub probe: Vec<Connection>,
    pub stale: Vec<Connection>,
}

/// Pure acknowledgement state for heartbeat adapters.
#[derive(Clone, Debug, Default)]
pub struct LivenessTracker<Connection> {
    alive: BTreeMap<Connection, bool>,
}

impl<Connection> LivenessTracker<Connection>
where
    Connection: Clone + Ord,
{
    #[must_use]
    pub fn new() -> Self {
        Self {
            alive: BTreeMap::new(),
        }
    }

    pub fn track(&mut self, connection: Connection) -> bool {
        if self.alive.contains_key(&connection) {
            return false;
        }
        self.alive.insert(connection, true);
        true
    }

    pub fn acknowledge(&mut self, connection: &Connection) -> bool {
        let Some(alive) = self.alive.get_mut(connection) else {
            return false;
        };
        *alive = true;
        true
    }

    pub fn remove(&mut self, connection: &Connection) -> bool {
        self.alive.remove(connection).is_some()
    }

    pub fn sweep(&mut self) -> LivenessSweep<Connection> {
        let mut probe = Vec::new();
        let mut stale = Vec::new();
        for (connection, alive) in &mut self.alive {
            if *alive {
                probe.push(connection.clone());
                *alive = false;
            } else {
                stale.push(connection.clone());
            }
        }
        for connection in &stale {
            self.alive.remove(connection);
        }
        LivenessSweep { probe, stale }
    }

    #[must_use]
    pub fn len(&self) -> usize {
        self.alive.len()
    }

    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.alive.is_empty()
    }

    pub fn clear(&mut self) -> usize {
        let removed = self.alive.len();
        self.alive.clear();
        removed
    }
}
